File: app/src/main/java/dev/ragnarok/fenrir/longpoll/GroupLongpoll.java
package dev.ragnarok.fenrir.longpoll;

import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import androidx.annotation.NonNull;

import java.util.concurrent.TimeUnit;

import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.api.interfaces.INetworker;
import dev.ragnarok.fenrir.api.model.longpoll.VkApiGroupLongpollUpdates;
import dev.ragnarok.fenrir.api.model.response.GroupLongpollServer;
import dev.ragnarok.fenrir.util.PersistentLogger;
import dev.ragnarok.fenrir.util.RxUtils;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

class GroupLongpoll implements ILongpoll {

    private static final int DELAY_ON_ERROR = 10 * 1000;

    private final int groupId;
    private final INetworker networker;
    private final Callback callback;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final Observable<Long> delayedObservable = Observable.interval(DELAY_ON_ERROR, DELAY_ON_ERROR,
            TimeUnit.MILLISECONDS, Injection.provideMainThreadScheduler());
    private String key;
    private String server;
    private String ts;

    GroupLongpoll(INetworker networker, int groupId, Callback callback) {
        this.groupId = groupId;
        this.callback = callback;
        this.networker = networker;
    }

    @Override
    public int getAccountId() {
        return -groupId;
    }

    private void resetServerAttrs() {
        server = null;
        key = null;
        ts = null;
    }

    @Override
    public void shutdown() {
        compositeDisposable.dispose();
    }

    @Override
    public void connect() {
        if (!isListeningNow()) {
            get();
        }
    }

    private boolean isListeningNow() {
        return compositeDisposable.size() > 0;
    }

    private void onServerInfoReceived(GroupLongpollServer info) {
        ts = info.ts;
        key = info.key;
        server = info.server;

        get();
    }

    private void onServerGetError(Throwable throwable) {
        PersistentLogger.logThrowable("Longpoll, ServerGet", throwable);
        getWithDelay();
    }

    private void get() {
        compositeDisposable.clear();

        boolean validServer = nonEmpty(server) && nonEmpty(key) && nonEmpty(ts);
        if (validServer) {
            compositeDisposable.add(networker.longpoll()
                    .getGroupUpdates(server, key, ts, 25)
                    .compose(RxUtils.applySingleIOToMainSchedulers())
                    .subscribe(this::onUpdates, this::onUpdatesGetError));
        } else {
            compositeDisposable.add(networker.vkDefault(getAccountId())
                    .groups()
                    .getLongPollServer(groupId)
                    .compose(RxUtils.applySingleIOToMainSchedulers())
                    .subscribe(this::onServerInfoReceived, this::onServerGetError));
        }
    }

    private void onUpdates(VkApiGroupLongpollUpdates updates) {
        if (updates.failed > 0) {
            resetServerAttrs();
            getWithDelay();
        } else {
            ts = updates.ts;

            if (updates.getCount() > 0) {
                callback.onUpdates(groupId, updates);
            }

            get();
        }
    }

    private void onUpdatesGetError(Throwable throwable) {
        PersistentLogger.logThrowable("Longpoll, UpdatesGet", throwable);
        getWithDelay();
    }

    private void getWithDelay() {
        compositeDisposable.add(delayedObservable.subscribe(o -> get()));
    }

    public interface Callback {
        void onUpdates(int groupId, @NonNull VkApiGroupLongpollUpdates updates);
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/longpoll/GroupLongpoll.kt
package dev.ragnarok.fenrir.longpoll

import dev.ragnarok.fenrir.Includes.provideMainThreadScheduler
import dev.ragnarok.fenrir.api.interfaces.INetworker
import dev.ragnarok.fenrir.api.model.longpoll.VkApiGroupLongpollUpdates
import dev.ragnarok.fenrir.api.model.response.GroupLongpollServer
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.util.PersistentLogger.logThrowable
import dev.ragnarok.fenrir.util.RxUtils.applySingleIOToMainSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.disposables.CompositeDisposable
import java.util.concurrent.TimeUnit

internal class GroupLongpoll(
    private val networker: INetworker,
    private val groupId: Int,
    private val callback: Callback
) : ILongpoll {
    private val compositeDisposable = CompositeDisposable()
    private val delayedObservable = Observable.interval(
        DELAY_ON_ERROR.toLong(), DELAY_ON_ERROR.toLong(),
        TimeUnit.MILLISECONDS, provideMainThreadScheduler()
    )
    private var key: String? = null
    private var server: String? = null
    private var ts: String? = null
    override val accountId: Int
        get() = -groupId

    private fun resetServerAttrs() {
        server = null
        key = null
        ts = null
    }

    override fun shutdown() {
        compositeDisposable.dispose()
    }

    override fun connect() {
        if (!isListeningNow) {
            get()
        }
    }

    private val isListeningNow: Boolean
        get() = compositeDisposable.size() > 0

    private fun onServerInfoReceived(info: GroupLongpollServer) {
        ts = info.ts
        key = info.key
        server = info.server
        get()
    }

    private fun onServerGetError(throwable: Throwable) {
        logThrowable("Longpoll, ServerGet", throwable)
        withDelay
    }

    private fun get() {
        compositeDisposable.clear()
        val validServer = server.nonNullNoEmpty() && key.nonNullNoEmpty() && ts.nonNullNoEmpty()
        if (validServer) {
            compositeDisposable.add(
                networker.longpoll()
                    .getGroupUpdates(server, key, ts, 25)
                    .compose(applySingleIOToMainSchedulers())
                    .subscribe({ updates: VkApiGroupLongpollUpdates -> onUpdates(updates) }) { throwable: Throwable ->
                        onUpdatesGetError(
                            throwable
                        )
                    })
        } else {
            compositeDisposable.add(
                networker.vkDefault(accountId)
                    .groups()
                    .getLongPollServer(groupId)
                    .compose(applySingleIOToMainSchedulers())
                    .subscribe({ info: GroupLongpollServer -> onServerInfoReceived(info) }) { throwable: Throwable ->
                        onServerGetError(
                            throwable
                        )
                    })
        }
    }

    private fun onUpdates(updates: VkApiGroupLongpollUpdates) {
        if (updates.failed > 0) {
            resetServerAttrs()
            withDelay
        } else {
            ts = updates.ts
            if (updates.count > 0) {
                callback.onUpdates(groupId, updates)
            }
            get()
        }
    }

    private fun onUpdatesGetError(throwable: Throwable) {
        logThrowable("Longpoll, UpdatesGet", throwable)
        withDelay
    }

    private val withDelay: Unit
        get() {
            compositeDisposable.add(delayedObservable.subscribe { get() })
        }

    interface Callback {
        fun onUpdates(groupId: Int, updates: VkApiGroupLongpollUpdates)
    }

    companion object {
        private const val DELAY_ON_ERROR = 10 * 1000
    }
}